ga4gh/server/frontend.py

```diff
@@ -729,23 +602,174 @@ def _parseTokenPayload(token):
+        """
+
+        access_map = {}
+
+        if app.config.get("TYK_ENABLED"):
+            if 'Authorization' in self.request_dict.headers:
+                access_map = {}
+                token = self.request_dict.headers['Authorization']
+                parsed_payload = _parseTokenPayload(token)
+
+                username = parsed_payload.get('preferred_username')
+                if username:
+                    access_map = app.access_map.getUserAccessMap().get(username, {})
+                else:
+                    if app.logger:
+                        app.logger.warn("Token does not contain a valid username")
             else:
-                prepare_counts[k] = [Counter(v)]
-
-    merged_counts = {}
-    for field in prepare_counts:
-        count_total = Counter()
-        for count in prepare_counts[field]:
-            count_total = count_total + count
-        merged_counts[field] = dict(count_total)
-    responseObject['results'][table] = [merged_counts]
+                raise exceptions.NotAuthenticatedException
+        else:
+            # for dev: mock full access when not using TYK config
+            token = None
+            for dataset in app.serverStatus.getDatasets():
+                access_map[dataset.getLocalId()] = 4
+
+        return token, access_map
+
+    def handleLocalRequest(self):
+        """
+        make local data request and set the results and status for a FederationResponse
+        """
+        try:
+            self.results = json.loads(
+                self.endpoint(
+                    self.request,
+                    return_mimetype=self.return_mimetype,
+                    access_map=self.access_map
+                )
+            )
+
+            self.status.append(200)
+
+        except (exceptions.ObjectWithIdNotFoundException, exceptions.NotFoundException):
+            self.status.append(404)
+
+        except (exceptions.NotAuthorizedException):
+            self.status.append(401)
+
+        print(">>> Local Response: " + str(self.status[0]))
+
+    def handlePeerRequest(self, request_type):
+        """
+        make peer data requests and update the results and status for a FederationResponse
+        """
+
+        header = {
+            'Content-Type': self.return_mimetype,
+            'Accept': self.return_mimetype,
+            'Federation': 'False',
+            'Authorization': self.token,
+        }
+
+        # generate peer uri
+        uri_list = []
+        for peer in app.serverStatus.getPeers():
+            uri = self.request_dict.url.replace(
+                self.request_dict.host_url,
+                peer.getUrl(),
+            )
+            uri_list.append(uri)
+
+        for future_response in self.async_requests(uri_list, request_type, header):
+            try:
+                response = future_response.result()
+            except (requests.exceptions.ConnectionError, requests.exceptions.Timeout):
```

Comment:
how long does it wait before it throws the timeout exception?

Reply:
Not sure, whatever our default server timeout is. We just catch the exception and keep going.